pvanalytics/quality/time.py

```diff
@@ -184,3 +184,103 @@ def shifts_ruptures(event_times, reference_times, period_min=2,
+        for shifts in `events`
+    min_different : int
+        Minimum difference between mean event time before `date` and mean
+        event time after `date` for a shift tom be detected.
+
+    Returns
+    -------
+    bool
+        Whether or not a shift of at least `min_difference` minutes was
+        detected in the the event times on `date`.
+    """
+    before = events[date - window:date - pd.Timedelta(days=1)]
+    after = events[date:date + window]
+    if len(before) == 0 or len(after) == 0:
+        raise ValueError(f"Insufficient data at {date}.")
+    before = before.dt.hour * 60 + before.dt.minute
+    after = after.dt.hour * 60 + after.dt.minute
+    return abs(before.mean() - after.mean()) > min_difference
+
+
+def has_dst(events, tz, window=7, min_difference=45):
+    """Return true if `events` appears to have daylight-savings shifts
+    at the dates in `shift_dates`.
+
+    Compares the mean event time in minutes since midnight over the
+    `window` days before and after each date in `shift_dates`. If the
+    difference is greater than `min_difference` then a shift has occurred
+    on that date.
+
+    Parameters
+    ----------
+    events : Series
+        Series with one timestamp for each day. The timestamp should
+        correspond to an event that occurs at roughly the same time on
+        each day, and shifts with daylight savings transitions. For example,
+        you may pass sunrise, sunset, or solar transit time.
+    tz : str
+        Name of a timezone that observes daylight savings and has the same
+        or similar UTC offset as the expected time zone for `events`.
+    window : int, default 7
+        Number of days before and after the shift date to consider. When
+        passing rounded timestamps in `events` it may be necessary to
+        use a smaller window. [days]
+    min_difference : int, default 45
+        Minimum difference between the mean event time before the shift
+        date and the mean event time after the event time. If the difference
+        is greater than `min_difference` a shift has occurred on that date.
+        [minutes]
+
+    Returns
+    -------
+    Series
+        Boolean Series indexed by dates on which Daylight Savings transitions
+        are expected to occur in `tz` and True if a shift was found on that
+        date in `events`
+
+    Raises
+    ------
+    ValueError
+        If there is no data before or after a shift date or there are no
+        daylight-savings shifts in `tz` for the dates covered by
+        `events`.
+    """
+    # Build a timestamp at noon on each day in the data
+    s = pd.Series(
+        pd.DatetimeIndex(events.index.tz_localize(None).date),
+        index=events.index
+    )
+    s = s + pd.Timedelta(hours=12)
+    s = s.dt.tz_localize(tz)
+    dst_shift = s.apply(lambda t: t.tzinfo.dst(t).total_seconds() / 3600)
+    shift_dates = s[dst_shift.diff().fillna(0) != 0]
```

Comment:
Factor out to its own function: given an index and a timezone return a series (or list?) with dates of the daylight savings transitions.